### PR TITLE
fix RpcInvokeContext not thread safe

### PR DESCRIPTION
**Pull request by OrezzerO**
### Motivation:
RpcInvokeContext is not thread safe , I add ReadWriteLock in it to make it thread safe

### Modification:

Describe the idea and modifications you've done.

### Result:

Fixes #<GitHub issue number>. 

If there is no issue then describe the changes introduced by this PR.

### CODE REVIEW

File: core/api/src/main/java/com/alipay/sofa/rpc/context/RpcInvokeContext.java [hunk 1]
```diff
@@ -172,7 +209,13 @@ public RpcInvokeContext setTimeout(Integer timeout) {
      */
     public void put(String key, Object value) {
         if (key != null && value != null) {
-            map.put(key, value);
+            Lock lock = rwlock.readLock();
```
Comment (JervyShi) [inline core/api/src/main/java/com/alipay/sofa/rpc/context/RpcInvokeContext.java:212]: Shouldn't we use write lock here?

File: core/api/src/main/java/com/alipay/sofa/rpc/context/RpcInvokeContext.java [hunk 2]
```diff
@@ -197,7 +240,13 @@ public Object get(String key) {
      */
     public Object remove(String key) {
         if (key != null) {
-            return map.remove(key);
+            Lock lock = rwlock.readLock();
```
Comment (JervyShi) [inline core/api/src/main/java/com/alipay/sofa/rpc/context/RpcInvokeContext.java:243]: Shouldn't we use write lock here? +1

File: core/api/src/main/java/com/alipay/sofa/rpc/context/RpcInvokeContext.java [hunk 3]
```diff
@@ -210,7 +259,13 @@ public Object remove(String key) {
      */
     public void putRequestBaggage(String key, String value) {
         if (BAGGAGE_ENABLE && key != null && value != null) {
-            requestBaggage.put(key, value);
+            Lock lock = rwlock.readLock();
```
Comment (JervyShi) [inline core/api/src/main/java/com/alipay/sofa/rpc/context/RpcInvokeContext.java:262]: Shouldn't we use write lock here? +1

File: core/api/src/main/java/com/alipay/sofa/rpc/context/RpcInvokeContext.java [hunk 4]
```diff
@@ -235,7 +290,13 @@ public String getRequestBaggage(String key) {
      */
     public String removeRequestBaggage(String key) {
         if (BAGGAGE_ENABLE && key != null) {
-            return requestBaggage.remove(key);
+            Lock lock = rwlock.readLock();
```
Comment (JervyShi) [inline core/api/src/main/java/com/alipay/sofa/rpc/context/RpcInvokeContext.java:293]: Shouldn't we use write lock here? +1

File: core/api/src/main/java/com/alipay/sofa/rpc/context/RpcInvokeContext.java [hunk 5]
```diff
@@ -246,17 +307,29 @@ public String removeRequestBaggage(String key) {
      * @return 全部响应透传数据
      */
     public Map<String, String> getAllRequestBaggage() {
-        return requestBaggage;
+        Lock lock = rwlock.readLock();
+        lock.lock();
+        try{
+            return requestBaggage;
```
Comment (JervyShi) [inline core/api/src/main/java/com/alipay/sofa/rpc/context/RpcInvokeContext.java:313]: Here, after the `readLock` is locked, the original map is returned directly, and the protection of the `writeLock` is broken when the map is traversed by an external party.

**Review by JervyShi — CHANGES_REQUESTED** [review]

File: core/api/src/main/java/com/alipay/sofa/rpc/context/RpcInvokeContext.java (hunk 5 above)
Comment (OrezzerO) [reply core/api/src/main/java/com/alipay/sofa/rpc/context/RpcInvokeContext.java]: You are right. I transform all the maps in `RpcInvokeContext` to ConcurrentHashMap, so they will not throw `ConcurrentModificationException`.

**Review by JervyShi — APPROVED** [review]
LGTM

File: core/api/src/main/java/com/alipay/sofa/rpc/context/RpcInvokeContext.java [hunk 6]
```diff
@@ -59,6 +96,45 @@ public static RpcInvokeContext getContext() {
...
+        //responseCallback
+        child.setResponseCallback(parent.getResponseCallback());
+        //future
+        child.setFuture(parent.getFuture());
+        //map
+        child.map.putAll(parent.map);
+        //customHeader
+        child.customHeader.putAll(parent.customHeader);
+
+        if (BAGGAGE_ENABLE) {
```
Comment (EvenLjj) [inline core/api/src/main/java/com/alipay/sofa/rpc/context/RpcInvokeContext.java:128]: The clone method only needs to copy according to the parent. Should there judge according to the BAGGAGE_ENABLE.

**Review by EvenLjj — APPROVED** [review]
LGTM

**Pull request merged**